packages/ethereum/contracts/src/Crypto.sol

```diff
@@ -0,0 +1,598 @@
+
+    (bool is_square, uint256 y1) = sqrt_ratio_3mod4(tv2, tv6); // 18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
+
+    uint256 y = mulmod(tv1, u, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 19.   y = tv1 * u
+    y = mulmod(y, y1, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 20.   y = y * y1
+
+    if (is_square) {
+      x = tv3; // 21.   x = CMOV(x, tv3, is_gx1_square)
+      y = y1; // 22.   y = CMOV(y, y1, is_gx1_square)
+    }
+
+    if (u % 2 != y % 2) { // 23.  e1 = sgn0(u) == sgn0(y)
+      y = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F - y; // 24.   y = CMOV(-y, y, e1)
+    }
+
+    r.x = mulmod(x, invMod(tv4), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 25.   x = x / tv4
+    r.y = y;
+  }
+
+  /**
+   * Square root subroutine used by simplified SWU mapping.
+   *
+   * Returns (true, sqrt(u / v)) if (u / v) is square in the field
+   * Returns (false, sqrt(Z * (u / v))) otherwise
+   *
+   * Optimizations:
+   * - mathematical optimization: reduce expmod / mulmod / addmod operations
+   * - few temporary values to reduce memory expansion
+   * - inlined constants to reduce memory expansion and prevent MLOAD instructions
+   */
+  function sqrt_ratio_3mod4(uint256 u, uint256 v) public view returns (bool, uint256) {
+    // uint256 C1 = 0x3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0b;
+    // uint256 C2 = 0x31fdf302724013e57ad13fb38f842afeec184f00a74789dd286729c8303c4a59;
+    // uint256 SECP256K1_BASE_FIELD_ORDER = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F;
+
+    uint256 tv1 = mulmod(v,v, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 1. tv1 = v^2
+    uint256 tv2 = mulmod(u,v, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 2. tv2 = u * v
+    tv1 = mulmod(tv1, tv2, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 3. tv1 = tv1 * tv2
+    uint256 y1 = expmod(tv1, 0x3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0b, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 4. y1 = tv1^c1
+    y1 = mulmod(y1, tv2, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 5. y1 = y1 * tv2
+    uint256 y2 = mulmod(y1, 0x31fdf302724013e57ad13fb38f842afeec184f00a74789dd286729c8303c4a59, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 6. y2 = y1 * c2
+    uint256 tv3 = mulmod(y1, y1, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 7. tv3 = y1^2
+    tv3 = mulmod(tv3, v, 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F); // 8. tv3 = tv3 * v
+
+    if (tv3 == u) { // 9. isQR = tv3 == u
+      return (true, y1);
+    } else {
+      return (false, y2);
+    } // 10. y = CMOV(y2, y1, isQR)
+  }
+
+  /**
+   * Consumes a byte-string and returns a field element of the secp256k1 base field.
+
```

Comment:
Docs should specify which hash (sha256) we're talking about.

Reply:
now part of the function name